### ISSUE TITLE
Migration to TypeScript?

### ISSUE THREAD

**Issue by bartsmykla** [kumahq/kuma-gui#119]
Our GUI is currently using Vue 2, and as we are days/maybe weeks before Vue 3 will become officially released as production ready, I want to start a discussion about moving from JS to TS (TypeScript).

I would love to know what you all think about this idea

My current thoughts:

Benefits:
- Types and all benefits related to having static typing
  - More confidence about what types we are providing to our components
  - Better integrations with IDEs
  - More future proof and easier to maintain code
  - Less bugs related to wrong types
- Vue 3 is written in TS

Drawbacks:
- I don't know how hard it would be to switch to use TS but it looks like there are tools to help with the process: https://cli.vuejs.org/core-plugins/typescript.html#installing-in-an-already-created-project and maybe it would be as smooth as just running `vue add typescript`
- (?) Some learning curve for contributors who doesn't know TS

**Comment by tharun208** [comment]
Yes, This will be an awesome improvement to make the code bulletproof 💯 . But, I see a practice in some typescript projects is that ppl start defining types with `any`, So, it will be easier for them to speed up their coding ( we need to avoid that ).

Also, it will be awesome if we even introduce `unit testing` using https://testing-library.com.

**Comment by bartsmykla** [comment]
@tharun208 thank you for a great suggestion. I would definitely look for not approving `any` as a type, also I'm working on introducing unit testing currently :-)

**Comment by darrenjennings** [comment]
@bartsmykla happy to help as we made this switch recently (and as the maintainers of Kongponents, we will of course be migrating it to be v3 compat and eventually TS as well!). My suggestions for your upgrade path:

1. add the vue-cli plugin https://cli.vuejs.org/core-plugins/typescript.html#vue-cli-plugin-typescript
  - https://cli.vuejs.org/core-plugins/typescript.html#installing-in-an-already-created-project
2. convert all .js to .ts
3. keep vue component script tags using default lang="js", and then begin switching components over to lang="ts"
4. For better or for worse, Vue 2's typescript story is incomplete. However, you can get the type safety you are looking for with the composition api https://github.com/vuejs/composition-api. I would install that and use `defineComponent` + `PropType<T>` which has been great.
5. Take this with a grain of salt, but I would go all in on strict mode if you can. We are not `strict` at the moment, and it will be painful to switch.
